src/parameters/matchparameter_nb.cpp

```diff
@@ -0,0 +1,83 @@
+#include "parameters/matchparameter_nb.h"
+
+#include "engine/api/match_parameters.hpp"
+
+#include <nanobind/nanobind.h>
+#include <nanobind/stl/string.h>
+#include <nanobind/stl/vector.h>
+
+namespace nb = nanobind;
+using namespace nb::literals;
+
+void init_MatchParameters(nb::module_& m) {
+    using osrm::engine::api::BaseParameters;
+    using osrm::engine::api::RouteParameters;
+    using osrm::engine::api::MatchParameters;
+
+    nb::enum_<MatchParameters::GapsType>(m, "MatchGapsType")
+        .value("Split", MatchParameters::GapsType::Split)
+        .value("Ignore", MatchParameters::GapsType::Ignore)
+        .export_values();
+
+    nb::class_<MatchParameters, RouteParameters>(m, "MatchParameters")
+        .def(nb::init<>())
+        .def("__init__", [](MatchParameters* t,
+            std::vector<unsigned> timestamps,
+            MatchParameters::GapsType gaps,
+            bool tidy,
+            std::vector<std::size_t> waypoints,
+                const bool steps,
+                const bool alternatives,
+                const RouteParameters::AnnotationsType annotations,
+                const RouteParameters::GeometriesType geometries,
+                const RouteParameters::OverviewType overview,
+                const boost::optional<bool> continue_straight,
+                    std::vector<osrm::util::Coordinate> coordinates,
+                    std::vector<boost::optional<osrm::engine::Hint>> hints,
+                    std::vector<boost::optional<double>> radiuses,
+                    std::vector<boost::optional<osrm::engine::Bearing>> bearings,
+                    std::vector<boost::optional<osrm::engine::Approach>> approaches,
+                    bool generate_hints,
+                    std::vector<std::string> exclude,
+                    const BaseParameters::SnappingType snapping
+            ) {
+                new (t) MatchParameters(
```

Comment:
I've done the MatchParameters in the "assigning" way, but I'm not sure if it would be clearer to just fully assign all values manually, or use the constructor for MatchParameters and RouteParameters, and then manually assign for BaseParameters. 

I believe the NodeJS bindings do it in the inverse way instead (ie. assigning BaseParameters, and then everything else is explicitly assigned), but I think that would be more complicated to replicate here.

Reply:
> but I'm not sure if it would be clearer to just fully assign all values manually

Yeah I agree here

Reply:
> you agree with manually assigning all values

exactly that:)

Reply:
you want to integrate that change before merging or make an issue out of it?

Reply:
After taking a closer look, it seems like default initializing MatchParameters and setting each parameter explicitly might result in additional complication. This is because when values are passed into the parent constructor, the parent constructor does perform some more transformations and assignments of its own.

For example, `AnnotationsType` gets converted to a boolean value, and the `number_of_alternatives` value is also calculated based off the passed `alternatives` value. As a result, I would have to copy over those calculations to the bindings side (ie. `number_of_alternatives = alternatives ? 1u : 0u`).

As a result, I think this approach may lose code clarity, since those value calculations would have to be replicated on the bindings side, and also require additional copied hardcoded values/calculations.

Reply:
Btw boolean `alternatives` is deprecated as well https://github.com/Project-OSRM/osrm-backend/blob/62e11fd5dbc8d5e888c6e2555505ae684d8a2759/include/engine/api/route_parameters.hpp#L168

Reply:
Ah, if it's `alternatives` bool input is getting deprecated, should we still support bool input for the Python bindings? 

The part that I didn't fully understand about the NodeJS `annotations_type` part there is that it seems to check what enum type was sent for `annotations` to assign the `annotations_type` value (in an array), but for the C++ side of things, it seems to base it off which RouteParameters constructor is called. If I'm understanding this part correctly, the user is able to send in an array of AnnotationsTypes, which gets calculated up? In that case, it would make sense to apply that same abstraction to the Python bindings.

Reply:
> Ah, if its alternatives bool input is getting deprecated, should we still support bool input for the Python bindings?

IMO no. Boolean never really made sense there and it should be int. 

The other stuff @SiarheiFedartsou is better suited answering:)

Reply:
> The part that I didn't fully understand about the NodeJS annotations_type part there is that it seems to check what enum type was sent for annotations to assign the annotations_type value (in an array), but for the C++ side of things, it seems to base it off which RouteParameters constructor is called. If I'm understanding this part correctly, the user is able to send in an array of AnnotationsTypes, which gets calculated up? In that case, it would make sense to apply that same abstraction to the Python bindings.

Not sure I fully understand the question... What happens in NodeJS bindings it simply converts something like this on JS side `["nodes", "distance", "speed"]` to smth like this on C++ side `std::vector<AnnotationType>{AnnotationType::Nodes, AnnotationType::Distance, AnnotationType::Speed}`. Can we do the same on Python side? Ofc we can use https://docs.python.org/3/library/enum.html or smth like this if needed.

Reply:
ahhh, I got it. `annotations_type` is not vector indeed, it is [bitmask](https://en.wikipedia.org/wiki/Mask_(computing)). Then you indeed can follow NodeJS path and just transform Python's list to this bitmask just like NodeJS bindings do. 

 So I would rephrase what I said above:
 
 > Not sure I fully understand the question... What happens in NodeJS bindings it simply converts something like this on JS side `["nodes", "distance", "speed"]` to smth like this on C++ side `AnnotationType::Nodes | AnnotationType::Distance | AnnotationType::Speed`. Can we do the same on Python side? Ofc we can use https://docs.python.org/3/library/enum.html or smth like this if needed.

Reply:
Okay, that sounds doable, thanks for the clarifications!

Reply:
I've made adjustments to accommodate the requested changes. In the process, I've also discovered a bug in osrm-backend's RouteParameter's AnnotationsType `|=` operator overload that renders it nonfunctional, so I've also put in a [PR over there](https://github.com/Project-OSRM/osrm-backend/pull/6646) to address that.